### ISSUE TITLE
Inaccurate stacktraces from multithreaded programs

### ISSUE THREAD

**Issue by tomasv** [rbspy/rbspy#387]
Hey,

Attaching to a running Rails Puma process seems to no longer work correctly. The Puma internal waiting thread stacktrace is always chosen instead of the active one.

I've narrowed it down to `rbspy` 0.18.0 as the first instance this happens.

Steps to reproduce:
1. `rails new test-app`
2. `rails s` inside of it to start the server.
3. `while true; do curl -s http://localhost:3000/ >/dev/null; done` to generate some activity.
4. `sudo -E rbspy record -p <PID_OF_PUMA_PROCESS>`

Reproduced on: Ruby 3.1.3, 3.2.2.
3.3 isn't trivial to check as support for it only appeared in 0.18.0.

Expected (approximately, this is what is captured in versions previous to 0.18.0):
```
Time since start: 11s. Press Ctrl+C to stop.
Summary of profiling data so far:
% self  % total  name
 72.17    72.98  select [c function] - (unknown)
 15.89    15.89  [] - <internal:dir>:129
  1.35     1.35  mtime [c function] - (unknown)
  0.90     0.90  wait_readable [c function] - (unknown)
  0.72     0.72  sleep [c function] - (unknown)
  0.63     0.63  write [c function] - (unknown)
  0.36     0.36  to_s [c function] - (unknown)
  0.36     0.36  exist? [c function] - (unknown)
  0.27     0.27  close [c function] - (unknown)
  0.18     1.53  block in max_mtime - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/activesupport-7.1.3/lib/active_support/file_update_checker.rb:139
  0.18     0.45  sprintf [c function] - (unknown)
  0.18     0.18  wakeup [c function] - (unknown)
  0.18     0.18  store - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/activesupport-7.1.3/lib/active_support/execution_context.rb:50
  0.18     0.18  read_nonblock - <internal:io>:64
  0.18     0.18  file? [c function] - (unknown)
  0.18     0.18  exception [c function] - (unknown)
  0.18     0.18  === [c function] - (unknown)
  0.09    22.98  call - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/actionpack-7.1.3/lib/action_dispatch/middleware/remote_ip.rb:93
  0.09     7.90  each [c function] - (unknown)
  0.09     4.58  call - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/activerecord-7.1.3/lib/active_record/migration.rb:656
```
[2024-02-22-DLJrCXU841.raw.gz](https://github.com/rbspy/rbspy/files/14372869/2024-02-22-DLJrCXU841.raw.gz)

What I see in >=0.18.0 (the output is mostly completely static):
```
Time since start: 10s. Press Ctrl+C to stop.
Summary of profiling data so far:
% self  % total  name
100.00   100.00  join [c function] - (unknown)
  0.00   100.00  with_argv - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/railties-7.1.3/lib/rails/command.rb:152
  0.00   100.00  tap - <internal:kernel>:92
  0.00   100.00  start - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/railties-7.1.3/lib/rails/commands/server/server_command.rb:41
  0.00   100.00  start - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rackup-2.1.0/lib/rackup/server.rb:342
  0.00   100.00  run - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/thor-1.3.0/lib/thor/command.rb:38
  0.00   100.00  run - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/puma-6.4.2/lib/rack/handler/puma.rb:82
  0.00   100.00  run - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/puma-6.4.2/lib/puma/single.rb:67
  0.00   100.00  run - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/puma-6.4.2/lib/puma/launcher.rb:197
  0.00   100.00  require [c function] - (unknown)
  0.00   100.00  require - <internal:/Users/tomas/.rbenv/versions/3.2.2/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:164
  0.00   100.00  require - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bootsnap-1.18.3/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:34
  0.00   100.00  perform - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/railties-7.1.3/lib/rails/commands/server/server_command.rb:150
  0.00   100.00  perform - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/railties-7.1.3/lib/rails/command/base.rb:74
  0.00   100.00  invoke_command - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/thor-1.3.0/lib/thor/invocation.rb:129
  0.00   100.00  invoke_command - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/railties-7.1.3/lib/rails/command/base.rb:181
  0.00   100.00  invoke - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/railties-7.1.3/lib/rails/command.rb:83
  0.00   100.00  dispatch - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/thor-1.3.0/lib/thor.rb:528
  0.00   100.00  block in perform - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/railties-7.1.3/lib/rails/commands/server/server_command.rb:149
  0.00   100.00  block in invoke - /Users/tomas/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/railties-7.1.3/lib/rails/command.rb:75
```
[2024-02-22-f10Vfk001c.raw.gz](https://github.com/rbspy/rbspy/files/14372867/2024-02-22-f10Vfk001c.raw.gz)

`rbspy` does work correctly for other processes, for example attaching to a running Rails console or a simple Ruby process. It seems something is not right with multithreaded applications.

Hope this helps 🙇

**Comment by acj** [comment]
Hey, thanks for the detailed report. I think the problem is in 0ce5089b5c82fc5de65df07fb09bf47e609a6bb6, which is much simpler but maybe too precise about finding the current execution context. If you're comfortable building rbspy from source, could you try reverting that commit and profiling Puma again? That seems to fix the problem for me on 3.2.3 (I'm using macOS). It can't get a stacktrace on 3.3.0, though, so I'll need to investigate that.

> 3.3 isn't trivial to check as support for it only appeared in 0.18.0.

To test newer versions, you can try `rbspy record --force-version 3.3.0 ...`. That probably won't work with 3.3.0 because of memory layout changes, but for most version bumps it should work.

**Comment by tomasv** [comment]
Something odd is happening though if I attempt to profile a Ruby 3.3 process with the revert, it does not display any stacktrace at all and fails with:
```
Time since start: 3s. Press Ctrl+C to stop.
Summary of profiling data so far:
% self  % total  name
Warning: no profile samples were collected
Wrote raw data to /Users/tomas/Library/Caches/rbspy/2024-02-23-urrlH7IOCo.raw.gz
Wrote formatted output to /Users/tomas/Library/Caches/rbspy/2024-02-23-ECJfr7jcKP.flamegraph.svg
Something went wrong while rbspy was sampling the process. Here's what we know:
- new spy
- Copy error: stack base and cfp address out of sync. stack base: ffffffc9, cfp address: 104bb2998
```
[2024-02-23-urrlH7IOCo.raw.gz](https://github.com/rbspy/rbspy/files/14384707/2024-02-23-urrlH7IOCo.raw.gz)

**Comment by acj** [comment]
Thanks for testing. Yeah, Ruby 3.3 had some changes related to threads and ractors that might be confusing rbspy. Our method of getting the execution context on 3.x (before 0ce5089b5c82fc5de65df07fb09bf47e609a6bb6) is a bit hacky but has worked well until now. I'll look into it.

**Comment by acj** [comment]
Okay, I think this is fixed on the main branch. Please give it a try when you have time.

**Comment by tomasv** [comment]
This is great, it works! Thank you. 🙇

**Comment by acj** [comment]
Released in [0.19.1](https://github.com/rbspy/rbspy/releases/tag/v0.19.1).
